From 45a4979291165bda8cd8e2fc2a14a0b23f8aa7ac Mon Sep 17 00:00:00 2001
From: Balazs Scheidler <bazsi@balabit.hu>
Date: Wed, 5 Jun 2013 10:51:23 +0200
Subject: [PATCH] logwriter: fixed an infinite loop in LogWriter

commit 45a4979291165bda8cd8e2fc2a14a0b23f8aa7ac upstream

This patch fixes a possible infinite loop, when the template expansion
results in a zero-length string. In that case syslog-ng assumed
it couldn't handle the message, sets consumed to FALSE, and puts
the original message back to the head of the queue.

This is a bad idea, as the next time the LogWriter wakes up, it finds
the same message there, does the same thing again and again ad infinum.

Signed-off-by: Balazs Scheidler <bazsi@balabit.hu>
---
 lib/logwriter.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/lib/logwriter.c b/lib/logwriter.c
index 97cf6c0..0be6a90 100644
--- a/lib/logwriter.c
+++ b/lib/logwriter.c
@@ -1022,6 +1022,12 @@ log_writer_flush(LogWriter *self, LogWriterFlushMode flush_mode)
               self->line_buffer->len = 0;
             }
         }
+      else
+        {
+          msg_debug("Error posting log message as template() output resulted in an empty string, skipping message",
+                    NULL);
+          consumed = TRUE;
+        }
       if (consumed)
         {
           if (lm->flags & LF_LOCAL)
-- 
1.7.0.2

